00_basics/sc_dimensions.ipynb

```diff
@@ -0,0 +1,771 @@
+     "text": [
+      "-> Set[ 3, 2, 1 ]"
+     ]
+    }
+   ],
+   "source": [
+    "a = Set[1, 2, 3, 1];\n",
+    "a;"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "upset-diary",
+   "metadata": {},
+   "source": [
+    "Notice that although we declared duplicates in the set this duplication is ignored by SuperCollider as a set can only have unique elements.\n",
+    "We can also quickly take a look at the cardinality of sets."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "id": "intimate-congress",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "-> 3"
+     ]
+    }
+   ],
+   "source": [
+    "a.size;"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "hawaiian-poverty",
+   "metadata": {},
+   "source": [
+    "Note also that due the circumstance that a set has no order we can also not access the *i*-th element of a set."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "id": "generic-consensus",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERROR: Message 'at' not understood.\n",
+      "RECEIVER:\n",
+      "Instance of Set {    (0x7f98d8263488, gc=C0, fmt=00, flg=00, set=02)\n",
+      "  instance variables [2]\n",
+      "    array : instance of Array (0x7f99101a8108, size=8, set=3)\n",
+      "    size : Integer 3\n",
+      "}\n",
+      "ARGS:\n",
+      "   Integer 0\n",
+      "\n",
+      "PROTECTED CALL STACK:\n",
+      "\tMeta_MethodError:new\t0x7f99183ac080\n",
+      "\t\targ this = DoesNotUnderstandError\n",
+      "\t\targ what = nil\n",
+      "\t\targ receiver = Set[ 3, 2, 1 ]\n",
+      "\tMeta_DoesNotUnderstandError:new\t0x7f99183ae040\n",
+      "\t\targ this = DoesNotUnderstandError\n",
+      "\t\targ receiver = Set[ 3, 2, 1 ]\n",
+      "\t\targ selector = at\n",
+      "\t\targ args = [ 0 ]\n",
+      "\tObject:doesNotUnderstand\t0x7f99181fb200\n",
+      "\t\targ this = Set[ 3, 2, 1 ]\n",
+      "\t\targ selector = at\n",
+      "\t\targ args = nil\n",
+      "\ta FunctionDef\t0x7f98d80f6a08\n",
+      "\t\tsourceCode = \"{ var result; \\\"**** JUPYTER ****\\\".postln; result = {a[0]}.value(); postf(\\\"-> %\n",
+      "\\\", result); \\\"\\\".postln;}\"\n"
+     ]
+    }
+   ],
+   "source": [
+    "a[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "blocked-weight",
+   "metadata": {},
+   "source": [
+    "## Working with matrices\n",
+    "\n",
+    "Lets start by implementing a function which will multiply two matrices which also checks if this is a valid operation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "id": "gothic-thong",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "-> a Function"
+     ]
+    }
+   ],
+   "source": [
+    "~matrixMul = {|a, b|\n",
```

Comment:
hm? `b.collect({|z| z})` that is the same as writing `b`.

In the seminat we have:

```supercollider
{ |a, b|
	b.collect { |coefs| sum(a * coefs) }
};
````

or safer:

```supercollider
{ |a, b|
	if(a.first.size != b.size) { Error("Matrices do not match in size")throw };
	b.collect { |coefs| sum(a * coefs) }
};
````

Reply:
Good point - regarding the `m` implementation see below

Reply:
Sorry, for this deconvoluted way we need to convert everything to column vecors and then back:

```
m = { |a, b|
	b = b.flop; a = a.flop;
	b.collect { |coefs| sum(a * coefs) }.flop
};
```

This is less efficient than stepwise iteration, but it makes more sense (to me). Well …